server/player/player.go

```diff
@@ -1050,6 +1064,70 @@ func (p *Player) StopFlying() {
 	p.session().SendGameMode(p.GameMode())
 }
 
+// Sleep makes the player sleep at the given position. If the position does not map to a bed (specifically the head side),
+// the player will not sleep.
+func (p *Player) Sleep(pos cube.Pos) {
+	w := p.World()
+	if b, ok := w.Block(pos).(block.Bed); ok {
+		if b.User != nil {
+			// The player cannot sleep here.
+			return
+		}
+		b.User = p
+		w.SetBlock(pos, b, nil)
+	}
+
+	w.SetRequiredSleepDuration(time.Second * 5)
```

Comment:
What do you mean by "something internal"? You must be sleeping in a bed for at least 5 seconds before the day progresses, and that cool down gets reset back to 5 every time someone else starts to sleep.

Reply:
I mean having only a required sleep duration in world's settings and updating/tracking how much longer must be slept inside of World automatically, instead of having a `SetRequiredSleepDuration`.

Reply:
That won't work though. The sleep duration is reset in vanilla if the player stops and then starts sleeping again. If you're suggesting we should make the default required sleep duration configurable via world settings, maybe, but that doesn't entirely make sense since vanilla doesn't support it anyway.